tower-ready-cache/src/service.rs

```diff
@@ -0,0 +1,380 @@
+            Some((idx, _, _)) => {
+                self.repair_next_ready_index(idx);
+                true
+            }
+        }
+    }
+
+    // Updates `next_ready_index_idx` after the entry at `rm_idx` was evicted.
+    fn repair_next_ready_index(&mut self, rm_idx: usize) {
+        self.next_ready_index = self.next_ready_index.take().and_then(|idx| {
+            let new_sz = self.ready.len();
+            debug_assert!(idx <= new_sz && rm_idx <= new_sz);
+            let repaired = match idx {
+                i if i == rm_idx => None,         // removed
+                i if i == new_sz => Some(rm_idx), // swapped
+                i => Some(i),                     // uneffected
+            };
+            trace!(
+                "repair_index: orig={}; rm={}; sz={}; => {:?}",
+                idx,
+                rm_idx,
+                new_sz,
+                repaired,
+            );
+            repaired
+        });
+    }
+}
+
+impl<K, S, Req> ReadyCache<K, S, Req>
+where
+    K: Clone + Eq + Hash,
+    S: Service<Req>,
+    <S as Service<Req>>::Error: Into<error::Error>,
+    S::Error: Into<error::Error>,
+{
+    /// Pushes a new service onto the unready set.
+    ///
+    /// The service will be promoted to the ready set as `poll_pending` is invoked.
+    ///
+    /// Note that this does not replace a service currently in the ready set with
+    /// an equivalent key until this service becomes ready or the existing
+    /// service becomes unready.
+    pub fn push_service(&mut self, key: K, svc: S) {
+        let (tx, cancel) = oneshot::channel();
+        if let Some(c) = self.cancelations.insert(key.clone(), tx) {
+            // If there is already a pending service for this key, cancel it.
+            c.send(()).expect("cancel receiver lost");
+        }
+        self.pending.push(EvictableReady {
+            key: Some(key),
+            ready: Ready::new(svc),
+            cancel,
+        });
+    }
+
+    /// Polls unready services to readiness.
+    ///
+    /// Returns `Async::Ready` when there are no remaining unready services.
+    /// `poll_ready` should be called again after `push_service` or
+    /// `call_ready_index` are invoked.
+    pub fn poll_pending(&mut self) -> Poll<(), error::Failed<K>> {
+        loop {
+            match self.pending.poll() {
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                Ok(Async::Ready(None)) => return Ok(Async::Ready(())),
+                Ok(Async::Ready(Some((key, svc)))) => {
+                    trace!("endpoint ready");
+                    self.cancelations.remove(&key).expect("cancel sender lost");
+                    self.ready.insert(key, svc);
+                }
+                Err(PendingError::Canceled(_)) => {
+                    debug!("endpoint canceled");
+                    // Either the cancelation has already been removed, or a
+                    // replacement unready service has already been pushed.
```

Comment:
I think you can make a stronger statement here, namely "The cancellation for this service was removed in order to cause this cancellation."